prime-router/src/main/kotlin/azure/PreviewFunction.kt

```diff
@@ -0,0 +1,191 @@
+            } catch (ex: Exception) {
+                logger.error("Internal error for ${it.userName} request: $ex")
+                HttpUtilities.internalErrorResponse(request)
+            }
+        }
+    }
+
+    /**
+     * Encapsulate the parameters of the functions.
+     */
+    data class FunctionParameters(
+        val previewMessage: PreviewMessage,
+        val receiver: Receiver,
+        val sender: Sender,
+    )
+
+    /**
+     * Throw an [IllegalArgumentException] with a message based on [message], [errors] and [warnings].
+     */
+    private fun badRequest(
+        message: String,
+        errors: List<ActionLog> = emptyList(),
+        warnings: List<ActionLog> = emptyList()
+    ): Nothing {
+        val previewErrorMessage = PreviewResponseMessage.Error(
+            message,
+            errors.map { it.toSummary() },
+            warnings.map { it.toSummary() }
+        )
+        val response = mapper.writeValueAsString(previewErrorMessage)
+        throw IllegalArgumentException(response)
+    }
+
+    /**
+     * Look at the request body.
+     * Check for valid values with defaults are assumed if not specified.
+     */
+    fun checkRequest(request: HttpRequestMessage<String?>): FunctionParameters {
+        fun lookupReceiver(organizations: List<DeepOrganization>?, receiverName: String): Receiver? {
+            if (organizations == null) return null
+            return FileSettings()
+                .loadOrganizationList(organizations)
+                .findReceiver(receiverName)
+        }
+
+        fun lookupSender(organizations: List<DeepOrganization>?, senderName: String): Sender? {
+            if (organizations == null) return null
+            return FileSettings()
+                .loadOrganizationList(organizations)
+                .findSender(senderName)
+        }
+
+        val body = request.body
+            ?: badRequest("Missing body")
+        val previewMessage = mapper.readValue(body, PreviewMessage::class.java)
+        val receiver = previewMessage.receiver
+            ?: lookupReceiver(previewMessage.deepOrganizations, previewMessage.receiverName)
+            ?: workflowEngine.settings.findReceiver(previewMessage.receiverName)
+            ?: badRequest("Missing receiver")
+        val sender = previewMessage.sender
+            ?: lookupSender(previewMessage.deepOrganizations, previewMessage.senderName)
+            ?: workflowEngine.settings.findSender(previewMessage.senderName)
+            ?: badRequest("Missing sender")
+        return FunctionParameters(previewMessage, receiver, sender)
+    }
+
+    /**
+     * Main logic of the Azure function. Useful for unit testing.
+     */
+    fun processRequest(parameters: FunctionParameters): PreviewResponseMessage.Success {
+        val warnings = mutableListOf<ActionLog>()
+        return readReport(parameters, warnings)
+            .translate(parameters, warnings)
+            .buildResponse(parameters, warnings)
```

Comment:
I think this pattern or something like it should be the goal for how we describe the steps a report or item goes though.

Reply:
Right. I'm trying to encapsulate the algorithm of the function in readable function calls.